### ISSUE TITLE
BUG: unreduced array is produced with custom-call and xmap manual sharding

### ISSUE THREAD

**Issue by seanprime7** [google/jax#11941]
### Description

I have a **custom-call** bwd operation that takes `x (4, 4, 8)`, `invvar (4, 1)`, `weight (4, 8)`, and `grad_output (4, 4, 8)`.  It produces `grad_input (4, 4, 8)`, `grad_weight (4, 8)`, and `scratch (16, 32)`.

I am running this bwd operation with `x`, `invvar`, `grad_output`, and `grad_input` sharded on the batch dimension (leftmost dim), and `weight`, `grad_weight`, and `scratch` replicated, using `xmap` manual sharding:

```python
jax.config.update("experimental_xmap_spmd_lowering", True)
jax.config.update("experimental_xmap_spmd_lowering_manual", True)
weight = jax.numpy.ones((4, 8), dtype=numpy.float32)
g = jax.random.normal(prng_key, (4, 4, 8))

out, res = jax_ext.fwd(x, weight)

jitted_out = jax.jit(
    partial(jax_ext._bwd_p.bind, eps=1e-05)
)(g, *res)
with Mesh(numpy.array(jax.devices()).reshape(4), ("a")):

    def xmap_bwd(res, g):
        invvar, x, weight = res
        invvar_sharded = invvar.reshape(4, 1)
        x_sharded = x.reshape(4, 4, 8)
        g_sharded = g.reshape(4, 4, 8)
        axes = ("a", None, None)
        xmapped = xmap(
            partial(
                jax_ext._bwd_p.bind, eps=1e-05
            ),
            in_axes=(axes, ("a", None), axes, (None, None)),
            out_axes=[axes, (None, None), (None, None)],
            axis_resources={"a": "a"},
        )
        xmapped_out = xmapped(g_sharded, invvar_sharded, x_sharded, weight)
        return xmapped_out

    pjitted = pjit(
        xmap_bwd,
        in_axis_resources=(
            (
                PartitionSpec(
                    None,
                ),
                PartitionSpec(None, None, None),
                PartitionSpec(None, None),
            ),
            PartitionSpec(None, None, None),
        ),
        out_axis_resources=[
            PartitionSpec(None, None, None),
            PartitionSpec(None, None),
            PartitionSpec(None, None),
        ],
    )
    print(jax.make_jaxpr(pjitted)(res, g))
    pjitted.lower(res, g).compiler_ir().dump()
    pjitted_out = pjitted(res, g)

print(f"jitted_out: {jitted_out}")
print(f"pjitted_out: {pjitted_out}")
print(numpy.allclose(jitted_out[0], pjitted_out[0]))
print(numpy.allclose(jitted_out[1], pjitted_out[1]))
print(numpy.allclose(jitted_out[2], pjitted_out[2]))
```

As `grad_weight` is a replicated array, each device gets its own buffer but they populate the buffers with different values, making it an **"unreduced"** array.  For my use case, the values in all buffers need to be summed to produce the correct final values for `grad_weight`.

What can I do to produce **"reduced"** `grad_input`?

### What jax/jaxlib version are you using?

jax/jaxlib 0.3.16 built on 08/11/2022 with TF cfc45e1027d43cf54b0358d7d9e2fc01f58938dd

### Which accelerator(s) are you using?

GPU

**Comment by apaszke** [comment]
Sorry I think that's not enough details for me to be able to solve that. I don't know what the semantics of the bwd op are, but in general xmap should insert a reduction for every named axis broadcast that happens in your computation.

**Comment by seanprime7** [comment]
@jekbradbury already helped us resolve this issue last Friday.
